akka-persistence/src/main/scala/akka/persistence/snapshot/local/LocalSnapshotStore.scala

```diff
@@ -126,9 +127,20 @@ private[persistence] class LocalSnapshotStore extends SnapshotStore with ActorLo
   private def snapshotMetadatas(persistenceId: String, criteria: SnapshotSelectionCriteria): immutable.Seq[SnapshotMetadata] = {
     val files = snapshotDir.listFiles(new SnapshotFilenameFilter(persistenceId))
     if (files eq null) Nil // if the dir was removed
-    else files.map(_.getName).collect {
-      case FilenamePattern(pid, snr, tms) ⇒ SnapshotMetadata(URLDecoder.decode(pid, UTF_8), snr.toLong, tms.toLong)
-    }.filter(md ⇒ criteria.matches(md) && !saving.contains(md)).toVector
+    else {
+      // optimised, since noticed to cause slowdowns when many snapshots present:
+      files.iterator.flatMap((file: File) ⇒ {
+        val empty = Iterable.empty[SnapshotMetadata] // since Iterable.empty goes through a new builder, we avoid this
+        extractMetadata(file.getName) match {
+          case Some((pid, snr, tms)) ⇒
+            val md = SnapshotMetadata(URLDecoder.decode(pid, UTF_8), snr, tms)
+            if (criteria.matches(md).&&(!saving.contains(md))) Iterable(md)
```

Comment:
Hah, will fix - artifact of me having used "explain scala" to see which conversions were applied and where on the array etc. This rewrite avoids converting Option as Iterable btw (previous one did)